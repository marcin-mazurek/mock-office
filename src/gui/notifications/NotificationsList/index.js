export { NotificationsList, NOTIFICATION_CLICKED, notificationsClickedAction } from './NotificationsList';
