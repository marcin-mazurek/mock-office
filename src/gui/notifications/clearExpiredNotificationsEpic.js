import { Observable } from 'rxjs';
import { NEW_NOTIFICATION_DISPLAYED } from './actions';

export const NOTIFICATION_EXPIRED = 'notifications/NOTIFICATION_EXPIRED';
export const notificationExpiredAction = id => ({
  type: NOTIFICATION_EXPIRED,
  id
});
const NOTIFICATION_EXPIRATION_TIME = 5000;
const expireNotification = ({ id }) => new Promise((resolve) => {
  setTimeout(() => resolve(id), NOTIFICATION_EXPIRATION_TIME);
});

export default action$ =>
  action$.ofType(NEW_NOTIFICATION_DISPLAYED)
    .flatMap(expireNotification)
    .map(id => notificationExpiredAction(id));
