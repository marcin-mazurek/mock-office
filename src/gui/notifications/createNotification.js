import { Record } from 'immutable';
import unique from 'cuid';

export const Notification = new Record({
  id: '',
  text: '',
  type: 'info'
});

export const createNotification = (params) => {
  const config = {};

  config.id = unique();
  switch (params.type) {
    case 'error': {
      config.type = 'error';
      break;
    }
    case 'success': {
      config.type = 'success';
      break;
    }
    default: {
      config.type = 'info';
    }
  }
  config.text = params.text;

  return new Notification(config);
};
