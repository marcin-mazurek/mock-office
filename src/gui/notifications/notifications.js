import { SUCCEED as ADD_SERVER_DID_SUCCEED, FAILED as ADD_SERVER_FAILED } from '../servers/addServer/epics';
import { DID_FAIL as EDIT_SERVER_DID_FAIL } from '../servers/editServer/epics';
import {
  DID_FAIL as REMOVE_SERVER_DID_FAIL,
  DID_SUCCEED as REMOVE_SERVER_DID_SUCCEED
} from '../servers/removeServer/epics';
import { DID_FAIL as REMOVE_MOCK_DID_FAIL } from '../mocks/removeMock/epics';
import {
  SUCCEEDED as IMPORT_MOCKS_SUCCEEDED,
  FAILED as IMPORT_MOCKS_FAILED
} from '../mocks/importMock/epics';
import {
  SUCCEED as ADD_HTTP_MOCK_SUCCEED,
  FAILED as ADD_HTTP_MOCK_FAILED
} from '../mocks/addMock/addHttpMock/epic';
import {
  SUCCEED as ADD_WS_MOCK_SUCCEED,
  FAILED as ADD_WS_MOCK_FAILED
} from '../mocks/addMock/addWsMock/epic';
import {
  FAILED as START_SERVER_FAILED
} from '../servers/startServer/epics';
import notificationsReduxModule from './notificationsReduxModule';

const showError = (state, action, module) =>
  module.addNotification(state, { type: 'error', text: action.reason });
const showAddMockSuccess = (state, action, module) =>
  module.addNotification(state, { type: 'success', text: 'Mock added' });

export const reducer = notificationsReduxModule.createReducer({
  [ADD_SERVER_DID_SUCCEED]: (state, action, module) =>
    module.addNotification(state, { type: 'success', text: 'Server added' }),
  [REMOVE_SERVER_DID_SUCCEED]: (state, action, module) =>
    module.addNotification(state, { type: 'success', text: 'Server removed' }),
  [EDIT_SERVER_DID_FAIL]: showError,
  [REMOVE_SERVER_DID_FAIL]: showError,
  [REMOVE_MOCK_DID_FAIL]: showError,
  [IMPORT_MOCKS_FAILED]: showError,
  [ADD_HTTP_MOCK_FAILED]: showError,
  [ADD_WS_MOCK_FAILED]: showError,
  [ADD_SERVER_FAILED]: showError,
  [START_SERVER_FAILED]: showError,
  [IMPORT_MOCKS_SUCCEEDED]: (state, action, module) =>
    module.addNotification(state, { type: 'success', text: 'Mocks imported' }),
  [ADD_HTTP_MOCK_SUCCEED]: showAddMockSuccess,
  [ADD_WS_MOCK_SUCCEED]: showAddMockSuccess
});
export const selectors = notificationsReduxModule.createSelectors(state => state.get('notifications'));
