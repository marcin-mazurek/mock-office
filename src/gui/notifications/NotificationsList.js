import React from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Notification } from './createNotification';
import { allNotificationsSelector } from './selectors';
import { removeAction as createRemoveAction } from './actions';

// eslint-disable-next-line import/prefer-default-export
export const NotificationsList = ({ notifications, remove }) => (
  <ul className="notifications-list">
    {
      notifications.map(
        ({ id, text, type }) =>
          <li className="notifications-list-item" key={id}>
            <button
              className={`notifications-bubble notifications-bubble--${type}`}
              onClick={() => remove(id)}
            >
              {text}
            </button>
          </li>
      )
    }
  </ul>
);
NotificationsList.propTypes = {
  notifications: ImmutablePropTypes.listOf(PropTypes.instanceOf(Notification)),
  remove: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  notifications: allNotificationsSelector(state)
});
const mapDispatchToProps = {
  remove: createRemoveAction
};

const NotificationsListConnect = connect(mapStateToProps, mapDispatchToProps)(NotificationsList);

export default NotificationsListConnect;
