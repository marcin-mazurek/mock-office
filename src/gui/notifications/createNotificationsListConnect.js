import { connect } from 'react-redux';
import { NotificationsList, notificationsClickedAction } from './NotificationsList';

const createNotificationsListConnect = (selectors) => {
  const mapStateToProps = state => ({
    notifications: selectors.allNotificationsSelector(state)
  });
  const mapDispatchToProps = {
    onNotificationClick: notificationsClickedAction
  };
  return connect(mapStateToProps, mapDispatchToProps)(NotificationsList);
};

export default createNotificationsListConnect;
