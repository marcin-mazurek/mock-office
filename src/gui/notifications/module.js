import { Map, List, Record } from 'immutable';
import { createSelector } from 'reselect';
import unique from 'cuid';
import { NOTIFICATION_CLICKED } from './NotificationsList';

export const Notification = new Record({
  id: '',
  text: '',
  type: 'info'
});

const createNotification = (params) => {
  const config = {};

  config.id = unique();
  switch (params.type) {
    case 'error': {
      config.type = 'error';
      break;
    }
    case 'success': {
      config.type = 'success';
      break;
    }
    default: {
      config.type = 'info';
    }
  }
  config.text = params.text;

  return new Notification(config);
};

const initialState = () => new Map({
  entities: new Map(),
  ids: new List()
});

const reducerActions = {
  [NOTIFICATION_CLICKED]: (action, reducers, state) =>
    reducers.removeNotification(state, action.id)
};

const moduleReducers = {
  addNotification(state, notificationParams) {
    const notification = createNotification(notificationParams);
    return state
      .update('entities', entities => entities.set(notification.id, notification))
      .update('ids', ids => ids.push(notification.id));
  },
  removeNotification(state, id) {
    return state
      .update('ids', ids => ids.filter(notificationId => notificationId !== id))
      .update('entities', entities => entities.delete(id));
  }
};

export const configureReducer = (actions) => {
  const allActions = Object.assign(reducerActions, actions);

  return (state = initialState(), action) => {
    const actionTypes = Object.keys(allActions);

    for (let i = 0; i < actionTypes.length; i += 1) {
      if (action.type === actionTypes[i]) {
        const transform = allActions[actionTypes[i]];
        return transform(action, moduleReducers, state);
      }
    }

    return state;
  };
};

export const configureSelectors = stateGetter => ({
  allNotificationsSelector: createSelector(
    state => stateGetter(state).get('ids'),
    state => stateGetter(state).get('entities'),
    (ids, entities) => ids.map(id => entities.get(id))
  )
});
