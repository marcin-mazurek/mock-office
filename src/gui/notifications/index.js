import {
  SUCCEEDED as ADD_SERVER_DID_SUCCEED,
  FAILED as ADD_SERVER_FAILED
} from '../servers/addServer/epics';
import { FAILED as EDIT_SERVER_DID_FAIL } from '../servers/editServer/epics';
import {
  FAILED as REMOVE_SERVER_DID_FAIL,
  SUCCEEDED as REMOVE_SERVER_DID_SUCCEED
} from '../servers/removeServer/epics';
import { DID_FAIL as REMOVE_MOCK_DID_FAIL } from '../mocks/removeMock/epics';
import {
  SUCCEEDED as IMPORT_MOCKS_SUCCEEDED,
  FAILED as IMPORT_MOCKS_FAILED
} from '../mocks/importMock/epics';
import {
  SUCCEED as ADD_HTTP_MOCK_SUCCEED,
  FAILED as ADD_HTTP_MOCK_FAILED
} from '../modules/addHttpMock/epic';
import {
  SUCCEED as ADD_WS_MOCK_SUCCEED,
  FAILED as ADD_WS_MOCK_FAILED
} from '../modules/addWsMock/epic';
import {
  FAILED as START_SERVER_FAILED
} from '../servers/startServer/startServer';
import { NOTIFICATION_CLICKED } from './NotificationsList';
import notificationsReduxModule from './notificationsReduxModule';
import createNotificationsListConnect from './createNotificationsListConnect';
import { FAILED as STOP_SERVER_FAILED } from '../servers/stopServer';
import { NOTIFICATION_EXPIRED } from './clearExpiredNotificationsEpic';

const showError = (state, action, reducers) =>
  reducers.addNotification(state, { type: 'error', text: action.reason });
const showAddMockSuccess = (state, action, reducers) =>
  reducers.addNotification(state, { type: 'success', text: 'Mock added' });

const notifications = notificationsReduxModule(
  {
    [ADD_SERVER_DID_SUCCEED]: (state, action, reducers) =>
      reducers.addNotification(state, { type: 'success', text: 'Server added' }),
    [REMOVE_SERVER_DID_SUCCEED]: (state, action, reducers) =>
      reducers.addNotification(state, { type: 'success', text: 'Server removed' }),
    [EDIT_SERVER_DID_FAIL]: showError,
    [REMOVE_SERVER_DID_FAIL]: showError,
    [REMOVE_MOCK_DID_FAIL]: showError,
    [IMPORT_MOCKS_FAILED]: showError,
    [ADD_HTTP_MOCK_FAILED]: showError,
    [ADD_WS_MOCK_FAILED]: showError,
    [ADD_SERVER_FAILED]: showError,
    [START_SERVER_FAILED]: showError,
    [STOP_SERVER_FAILED]: showError,
    [IMPORT_MOCKS_SUCCEEDED]: (state, action, reducers) =>
      reducers.addNotification(state, { type: 'success', text: 'Mocks imported' }),
    [ADD_HTTP_MOCK_SUCCEED]: showAddMockSuccess,
    [ADD_WS_MOCK_SUCCEED]: showAddMockSuccess,
    [NOTIFICATION_CLICKED]: (state, action, reducers) =>
      reducers.removeNotification(state, action.id),
    [NOTIFICATION_EXPIRED]: (state, action, reducers) =>
      reducers.removeNotification(state, action.id)
  },
  state => state.get('notifications')
);

export const reducer = notifications.reducer;
export const selectors = notifications.selectors;
export const NotificationsListConnect = createNotificationsListConnect(notifications.selectors);
