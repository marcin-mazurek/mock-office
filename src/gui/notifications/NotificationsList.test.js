import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { List } from 'immutable';
import { Provider } from 'react-redux';
import NotificationsListConnect, { NotificationsList } from './NotificationsList';
import { Notification } from '../entities/notifications/createNotification';
import configureStore from '../store';

jest.mock('node-unique', () => () => 'id');

describe('NotificationsList', () => {
  test('snapshot: with notifications', () => {
    const notifications = new List([
      new Notification({ text: 'Something happened' })
    ]);
    const wrapper = shallow(<NotificationsList notifications={notifications} remove={() => {}} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('snapshot: empty', () => {
    const notifications = new List([]);
    const wrapper = shallow(<NotificationsList notifications={notifications} remove={() => {}} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('NotificationsListConnect', () => {
  test('snapshot: connected to store with notifications', () => {
    const store = configureStore();

    store.dispatch({
      type: 'notifications/ADD',
      notification: {
        text: 'text'
      }
    });

    const wrapper = mount(
      <Provider store={store}>
        <NotificationsListConnect />
      </Provider>
    );

    expect(toJson(wrapper.find(NotificationsList))).toMatchSnapshot();
  });
});
