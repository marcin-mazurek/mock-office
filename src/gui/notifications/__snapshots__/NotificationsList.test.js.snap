// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsList snapshot: empty 1`] = `
<ul
  className="notifications-list"
/>
`;

exports[`NotificationsList snapshot: with notifications 1`] = `
<ul
  className="notifications-list"
>
  <li
    className="notifications-list-item"
  >
    <button
      className="notifications-bubble notifications-bubble--info"
      onClick={[Function]}
    >
      Something happened
    </button>
  </li>
</ul>
`;

exports[`NotificationsListConnect snapshot: connected to store with notifications 1`] = `
<Component
  notifications={
    Array [
      Object {
        "id": "id",
        "text": "text",
        "type": "info",
      },
    ]
  }
  remove={[Function]}
>
  <ul
    className="notifications-list"
  >
    <li
      className="notifications-list-item"
    >
      <button
        className="notifications-bubble notifications-bubble--info"
        onClick={[Function]}
      >
        text
      </button>
    </li>
  </ul>
</Component>
`;
